### ISSUE TITLE
Consistent Crashing on iPhone 6 (Does not produce logs)

### ISSUE THREAD

**Issue by cjnott1999** [kpwn/yalu102#206]
Upon jailbreak of 4k device (iPhone 7,2) consistent, and to my current knowledge, seemingly random crashing of what appears to be coreduetd along with several unknown crash logs. Attached is the crash log for coreduetd. 

[https://ghostbin.com/paste/g9uhx](url)

**Comment by AppleBetas** [comment]
Do you happen to have an Apple Watch connected?

**Comment by cjnott1999** [comment]
@AppleBetas Yes, yes I do. Are there known issues with apple watch compatibility? I saw something similar in another issue, but didn't seem to be exactly what I was experiencing.

**Comment by SparksCatcher** [comment]
Yes there are many issues with the apple watch. Crashing apps. Blank icons, Delayed notifications, and now this

**Comment by AppleBetas** [comment]
There are a few known issues, but those don’t appear to be as severe as this one. I also have an Apple Watch and iPhone 6, with coreduetd crashing. Since coreduetd does do a few things regarding Apple Watch, it could be partly because of it.

**Comment by cjnott1999** [comment]
I am attempting to replicate the issue and collect more logs now

**Comment by karboN90** [comment]
Not just with Apple Watch. 
I don't have a Apple Watch. And the problem still exists. 
Everything works fine but there is sudden crash and reboot of the iPhone

**Comment by cjnott1999** [comment]
I appears the crashes i am experiencing don't produce logs at all. Unknown crashes. all around

**Comment by tamer3030** [comment]
same here Iphone 6+ random restarts specially when I connect my Bluetooth headset

**Comment by silerh7** [comment]
Yalu crashes when I open safari or messages for a little while, it's somewhat stable everywhere else

**Comment by msamuel002** [comment]
i wonder if he fixed any crashes on the iPhone 6 in the newest commit (10 minutes ago). He didnt say anything about it but

**Comment by tamer3030** [comment]
I will compile the current commit and check

**Comment by kpwn** [comment]
Please test last commit. Closing this for now, reopen if issue persists.

**Comment by liudayu** [comment]
Issue persists. After I press GO in the app, most of the time device produces a "coreduetd" log then reboots, few times it resprings and back in jailbroken mode.

**Comment by nullpixel1** [comment]
@applebetas is also having it still happen

**Comment by kpwn** [comment]
oh i'm tripping. i thought this was my iPhone 6 is not stable issue. sorry, reopening. haven't looked into this

**Comment by nullpixel1** [comment]
Also, it looks like a permission issue, and I think it's linked to the watch issuee

**Comment by liudayu** [comment]
This is the coreduetd log file.
[Log.txt](https://github.com/kpwn/yalu102/files/737654/Log.txt)

**Comment by nullpixel1** [comment]
Ooo that looks like a permission issue even more now

**Comment by nullpixel1** [comment]
coreduetd handles stuff like watch connectivity and handoff for the record

**Comment by kpwn** [comment]
Is this only happening while in jailbroken mode?

**Comment by nullpixel1** [comment]
Yep

**Comment by liudayu** [comment]
No, when you reactivating your iPhone by pressing GO, sometimes it fails and then reboot, this is the only log it produces.

**Comment by nullpixel1** [comment]
I tried to start it manually, and it crashes almost instantly

**Comment by cjnott1999** [comment]
It appears from timing that coreduetd actually crashes before the phone itself reboots. Then it produces an unknown log.

**Comment by AppleBetas** [comment]
I also was just able to trigger a crash by switching to Safari on my Mac (therefore activating Handoff).

**Comment by bobbysn0w** [comment]
After the latest commit I have expierenced 0 crashes or bugs. I'll try and stress test it to see if it crashes.

**Comment by liudayu** [comment]
It does for me, just got a panic log and reset counter log after the latest commit

[panic-2017-01-30-024234.txt](https://github.com/kpwn/yalu102/files/737663/panic-2017-01-30-024234.txt)

**Comment by kpwn** [comment]
Ew, a SError with version set to MarijuanARM64. Are you sure you got the last commit?

**Comment by kpwn** [comment]
Anyway, that's a KPP failure. Had it happen due to inconsistent branch prediction. At least not a unknown memory access fault!

**Comment by liudayu** [comment]
Yes I did git-pull before I opened my Xcode. I will delete everything again and do a fresh clone, then see how it goes.

**Comment by karboN90** [comment]
@bobbysnow
You sure the latest commit experiences no crash at all ?

**Comment by cjnott1999** [comment]
Major kernel panic on clicking "go" with latest commit. In Xcode console the jailbreak doesn't reach completion. no logs generated.

**Comment by kpwn** [comment]
This is not 100% reliable, and expected. You should have a 50/50 chance of the exploit working.

**Comment by cjnott1999** [comment]
@kpwn Ah, still getting consistent crashing, again, no logs. Crashing happened quicker than last time upon reboot into a jailbroken state

**Comment by andrewwiik** [comment]
@kpwn rebooted, almost got to pressing the mail app before it crashed, iP6 10.2

**Comment by malhal** [comment]
After todays commits I successfully jailbroke my iPod Touch 6th Gen, thanks can now get on with updating my tweaks

**Comment by callum-99** [comment]
Also getting a random panic with the same panic log as @liudayu (iPhone 7,2 10.2)

**Comment by justinoboyle** [comment]
Latest commit randomly panics as well. i6.

Also, Cydia just crashes, but SSH works.

**Comment by msamuel002** [comment]
Yea at least luca knows what the issue is. Hope its an easy fix

**Comment by chashmeetsingh** [comment]
works on iPhone 6

**Comment by msamuel002** [comment]
@chashmeetsingh no crashing at all? how lol

**Comment by chashmeetsingh** [comment]
did once -> rejailbreak -> works fine now

**Comment by msamuel002** [comment]
hmm, i havent tried it out since they said its still not fixed. let me try

**Comment by RajPatel0929** [comment]
rejailbreak won't work/// rebooted 4 times and try it still does not work

**Comment by justinoboyle** [comment]
@chashmeetsingh Try doing a few things on your device, it will probably crash in a few seconds.

**Comment by chashmeetsingh** [comment]
hmm right did

**Comment by msamuel002** [comment]
ohh

**Comment by andrewwiik** [comment]
I think it is important to reboot inbetween a failed patching that causes crashes, is this correct? @kpwn

**Comment by msamuel002** [comment]
lol so basically luca just fixed the 5S and accidentally left the 6 undone? (which is fine idc)

**Comment by kpwn** [comment]
I am unable to repro this over multiple rejailbreaks.

**Comment by msamuel002** [comment]
what does that mean lol @kpwn

**Comment by msamuel002** [comment]
ohhh lol reproduce

**Comment by kpwn** [comment]
that this is not happening on my 6.

**Comment by msamuel002** [comment]
oh so this means it wont be easy to fix since its not happening? that sucks

**Comment by msamuel002** [comment]
Kim Jong Un has given your iPhone blessings

**Comment by chashmeetsingh** [comment]
it jailbroke successfully but after using it for like 5-10 mins crashed automatically @kpwn

**Comment by andrewwiik** [comment]
@kpwn I believe it may be because of people not rebooting inbetween random crashes, which is making the crashes worse everytime, I could be wrong though

**Comment by justinoboyle** [comment]
@kpwn Interesting. Do you have an Apple Watch paired (I do and am experiencing this). Latest commit.

I can't seem to find any error logs for some strange reason, but in a few seconds the phone will sporadicly crash. I went a few hours strong a few minutes ago, but now it just crashes ~30sec after the jailbreak.

**Comment by liudayu** [comment]
Yes, on high CPU usage (eg open snapchat, then twitter, facebook etc) will result a reboot

**Comment by msamuel002** [comment]
i agree with @liudayu

**Comment by kpwn** [comment]
I have ran a benchmark for an hour now.

**Comment by georgemcnish** [comment]
yeah, whenever I bring up multitasking it seems to crash or when I open an app it crashes, could it have something to do with corrupted memory?

**Comment by kpwn** [comment]
I wonder if I am by accident writing into KPP-checked memory.

**Comment by msamuel002** [comment]
@georgemcnish yea i think its memory overflow?

**Comment by andrewwiik** [comment]
@kpwn but don't you bypass KPP all together in the end?

**Comment by georgemcnish** [comment]
hmm

**Comment by liudayu** [comment]
Maybe try multiple apps, it will result either panic log or coreduetd & resetCounter logs

**Comment by msamuel002** [comment]
@liudayu would it come on the activate console on xcode?

**Comment by msamuel002** [comment]
or on diagnostics and usage?

**Comment by liudayu** [comment]
Those logs are in diagnostics and usage

**Comment by nullpixel1** [comment]
coreduetd handles handoff

Do you have handoff on @kpwn

**Comment by cjnott1999** [comment]
@kpwn I may have a different method of reproducing issue and getting logs. I will try and get back.  I feel it's a unique issue. Seemingly random.

**Comment by kpwn** [comment]
I do not, but I believe coreduetd issue is unrelated to KPP failures.

**Comment by msamuel002** [comment]
also, wayyyyyy off topic, but will there be any tool we can use to install this jailbreak instead of using cydia impactor (since the certificate lasts 7 days at a time)?

**Comment by kpwn** [comment]
For the record: I am running a forkbomb as well as several threads stuck in a tight syscall loop. This is the best benchmark for debugging KPP issues, and no crashes.

**Comment by andrewwiik** [comment]
I had it on, I just turned it off, I'll report back if it is more stable now with handoff disabled

**Comment by liudayu** [comment]
@msamuel002  Delete the off topic comment pls, and yes you can install from Xcode, free dev 7 days, and paid for 1 year.

**Comment by nullpixel1** [comment]
Try reproing it with handoff

**Comment by msamuel002** [comment]
handoff as in the feature where you can type on one device then transfer it to another? lol im dumb

**Comment by kpwn** [comment]
This is all on 10.2, right?

**Comment by msamuel002** [comment]
yep

**Comment by nullpixel1** [comment]
Indeed

**Comment by liudayu** [comment]
Yes, iOS 10.2

**Comment by tamer3030** [comment]
yes

**Comment by tomlube** [comment]
Yeah still crashing for me on iOS 10.2 iPhone 6. I do also have an Apple Watch as well...

**Comment by nullpixel1** [comment]
@andrewwiik

yep rip me too

**Comment by msamuel002** [comment]
i got a log in diagnostics and usage called "com.appl.WebKit.WebContent.cpu_resource-2017-01-29-105305.ips"....does this have to do with the jb

**Comment by kpwn** [comment]
WriteAnywhere32(kppsh+n, 0x580001e1); n+=4; // ldr	x1, #60
        WriteAnywhere32(kppsh+n, 0x58000140); n+=4; // ldr	x0, #40
        WriteAnywhere32(kppsh+n, 0xd5182020); n+=4; // msr	TTBR1_EL1, x0
        WriteAnywhere32(kppsh+n, 0xd2a00600); n+=4; // movz	x0, #0x30, lsl #16
        WriteAnywhere32(kppsh+n, 0xd5181040); n+=4; // msr	CPACR_EL1, x0
        WriteAnywhere32(kppsh+n, 0xd5182021); n+=4; // msr	TTBR1_EL1, x1
        WriteAnywhere32(kppsh+n, 0x10ffffe0); n+=4; // adr	x0, #-4
        WriteAnywhere32(kppsh+n, isvad ? 0xd5033b9f : 0xd503201f); n+=4; // nop
        WriteAnywhere32(kppsh+n, isvad ? 0xd508871f : 0xd508873e); n+=4; // tlbi vmalle1 (4k) / tlbi	vae1, x30 (16k)
        WriteAnywhere32(kppsh+n, 0xd5033fdf); n+=4; // isb
        WriteAnywhere32(kppsh+n, 0xd65f03c0); n+=4; // ret
        WriteAnywhere64(kppsh+n, ReadAnywhere64(ttbr0_real)); n+=8;
        WriteAnywhere64(kppsh+n, physp); n+=8;
        WriteAnywhere64(kppsh+n, physp); n+=8;

**Comment by kpwn** [comment]
Try using this as kpp shellcode.

**Comment by andrewwiik** [comment]
okay

**Comment by RajPatel0929** [comment]
@kpwn yes on 10.2

**Comment by cjnott1999** [comment]
ok. safari is a major issue i get consistent crashing with. finally got a log. @kpwn 
[log.txt](https://github.com/kpwn/yalu102/files/737680/log.txt)

**Comment by andrewwiik** [comment]
is this for coreduet or the other crashes btw? @kpwn

**Comment by msamuel002** [comment]
http://pastebin.com/Es74WqRy i think its safari crashing due to high cpu usage

**Comment by kpwn** [comment]
I believe the reason behind these issues is related to cache management, and Safari may trigger that due to JIT and stuff. But I dunno. Need to test.

**Comment by msamuel002** [comment]
and i also got 8 "panic-2017-01-29-xxxxxx.ips" logs but they are like SOOOO LONG

**Comment by nullpixel1** [comment]
Holy fuck

**Comment by nullpixel1** [comment]
That's some high cpu usage

**Comment by msamuel002** [comment]
ikr funny thing is that i havent even opened safari in like months @nullpixel1

**Comment by andrewwiik** [comment]
just ran, okay now what should I do? what should I open?

**Comment by nullpixel1** [comment]
Did the shellcode work?

**Comment by justinoboyle** [comment]
@andrewwiik try safari and opening a bunch of apps to get high CPU usage

**Comment by nullpixel1** [comment]
Try opening a page from handoff in safari on mac

**Comment by chashmeetsingh** [comment]
try updating cydia https://t.co/JhthOpc3dG

**Comment by georgemcnish** [comment]
Could it be WebKit treating the code as corrupted as safari could be running in the background? as stated in one of the bug reports
![image](https://cloud.githubusercontent.com/assets/19760795/22406393/56df894a-e64a-11e6-98e2-540933cd621a.png)

**Comment by andrewwiik** [comment]
@kpwn just opened a ton of shit, handoff pages in safari, a few social media apps, and everything seems to be good so far

**Comment by msamuel002** [comment]
yup @kpwn one of my panic logs says "SError". the pastebin for the shortest of those panic logs is here: http://pastebin.com/UMTjgd4k

**Comment by msamuel002** [comment]
dang that sounds good @andrewwiik

**Comment by nullpixel1** [comment]
Okay, I just borrowed someones 6 again, and that works fine! It never worked before

**Comment by andrei085** [comment]
My 6 is also crashing for the same reason, especialy when opening fb or big safari pages with a lot of images.

**Comment by cjnott1999** [comment]
@kpwn new app shell code seems to work so far.... will continue to brute test in jailbroken state

**Comment by justinoboyle** [comment]
Experiencing similar results to @andrewwiik . Try Cydia? It just crashes for me, but SSH works fine

**Comment by Jailbreakers** [comment]
Last commit 11 minutes ago. Please someone compile it and test it

**Comment by kpwn** [comment]
Note that this fix is not 'proper': it will have a perf impact due to more cache flushing. It shouldn't be too bad as long as your code doesn't spin on a syscall.

**Comment by kpwn** [comment]
The issue here is that this is subtle behaviour inherent to the CPU design, that may vary across different cores, and is entirely undocumented.

**Comment by kpwn** [comment]
The proper 'documented' way is guaranteed to work: but it's also guaranteed to be slow.

**Comment by nullpixel1** [comment]
Want me to PR? Or you got this?

Also, will it be helpful if I go through and review all the PRs? There's some good ones

**Comment by msamuel002** [comment]
what will happen if it spins on a syscall? boot loop?

**Comment by justinoboyle** [comment]
@Jailbreakers What do you think literally everyone here is doing?

**Comment by kpwn** [comment]
No: it will continuously call into KPP and will flush caches + entirety of tlb every time. Which is a performance hit.

**Comment by cjnott1999** [comment]
@kpwn lol. I have not had a crash since re-implementing the kpp shell. so the improper fix works for now.

**Comment by kpwn** [comment]
Note that this is intrinsic to my KPP bypass: a small slowdown is to be expected (note that this is a fixed amount of slowdown for each syscall: raw performance is still identical as long as no user->kernel transitions happen. when they do happen, kernel performance is identical up to the point where it has to return back to userland (and kpp runs), which with this fix includes a full cache and tlb flush, which is something that takes a bit of time).

This is also true for 16K devices, however I am able to not flush anything but 1 TLB entry without making the device unstable, which is *way* faster (almost within margin of error).

**Comment by kpwn** [comment]
I believe that this sort of performance hit can only be felt with targeted microbenchmarks. I can't think of a reason for someone to tightly loop into a syscall.

**Comment by cjnott1999** [comment]
@kpwn see what you're saying. this was never guaranteed to be stable, its a tradeoff between  performance (maybe?) over an occasional crash.

**Comment by kpwn** [comment]
Anyway, another few ACKs and I will commit the new shellcode and close this.

**Comment by nullpixel1** [comment]
@kpwn Am I good to add the 6 to supported devices table in #141?

**Comment by Dason-g** [comment]
@kpwn I can not install the tweaks. help me please

**Comment by beanjo55** [comment]
@Dason-g #64

**Comment by cjnott1999** [comment]
Issue to be fixed.

**Comment by kpwn** [comment]
Even if this is closed, I'd still like some more ACKs that it is indeed fixed, given this is subtle behaviour.

**Comment by nullpixel1** [comment]
Yeah. Can anyone confirm it works on the rest of the 4K devices? I'm about to update #141's table.

Or @kpwn, because #141 is so huge, should I make a new PR for readmes?

**Comment by andrewwiik** [comment]
No crash after running a shitty VR App I made, which leaves the CPU at 100%, consider it fixed for me

**Comment by KILLCAMPER** [comment]
Any iPad Air 2 cellular reports as I am getting no where fast

**Comment by kpwn** [comment]
iPad Air 2 is not supported yet.

**Comment by andrewwiik** [comment]
@kpwn is the iPad Air 2 the only device with 3 cores? I would of thought the iPad Pro would have a higher core count

**Comment by nullpixel1** [comment]
son:tm:

**Comment by liudayu** [comment]
@andrewwiik with handoff on or off?

**Comment by georgemcnish** [comment]
Running well for me, no crashes in where it used to crash

**Comment by nullpixel1** [comment]
on my gf's 6, I couldn't repro with handoff on or off with the new shellcode

**Comment by andrewwiik** [comment]
@liudayu hand off is on, I opened a few web pages with handoff as suggested too

**Comment by liudayu** [comment]
@andrewwiik Do you still experience sometimes you need to click on icon twice to open an app?

**Comment by msamuel002** [comment]
@kpwn have you fixed this in the latest commit yet? please let me know when you have :)

**Comment by msamuel002** [comment]
is it the "cacheshit"??? lmao

**Comment by andrewwiik** [comment]
@liudayu nope, first click it works

**Comment by andrewwiik** [comment]
nope, icons are fine

**Comment by KILLCAMPER** [comment]
What's the crap 💩 doing the rounds 
http://repo.hackyouriphone.org/jailbreak

**Comment by karboN90** [comment]
Since the lastest commit seem to solve the crash problem in i6 partly, 
Can someone compile the IPA and give us a download link to it ? 
Please 
Thank you

**Comment by nullpixel1** [comment]
It's now on Yalus beta site

**Comment by andrewwiik** [comment]
@kpwn still no crashes, started using FLEXLoader which enumerates through the entire heaps of apps for live debugging so I think it is safe to say we are good

**Comment by karboN90** [comment]
No it's not ?

**Comment by karboN90** [comment]
There is no official IPA on the Yalu webpage 🤗 
Can someone share the link of their own compiled IPA ? 
@andrewwiik ?

**Comment by Mila432** [comment]
@karboN90 delete your internet mate

**Comment by georgemcnish** [comment]
https://yalu.qwertyoruiop.com/yalu102_beta.ipa

**Comment by karboN90** [comment]
Just refreshed 
Got it 
Thank you 
@mila432 suck a dick

**Comment by msamuel002** [comment]
chill @karboN90 lmao

**Comment by msamuel002** [comment]
btw qwerty, when you said "a small slowdown is to be expected", does this mean battery life is gonna be worse. obviously with a jb it has to be worse, but would it be worse right now then a more stable jb version in the future?

**Comment by msamuel002** [comment]
nevermind i see exactly what you mean. it is alot slower for now but its fine, as long as we have a jailbreak :) thanks!

**Comment by andrewwiik** [comment]
@kpwn still no crashes :)
